hikari/impl/bot.py

```diff
@@ -23,589 +23,481 @@
+        )
 
-        This is a monotonic time. To get the physical startup date, see
-        `BotApp.started_at` instead.
+        for window_start in range(0, shard_count, requirements.session_start_limit.max_concurrency):
+            window = [
+                candidate_shard_id
+                for candidate_shard_id in range(
+                    window_start, window_start + requirements.session_start_limit.max_concurrency
+                )
+                if candidate_shard_id in shard_ids
+            ]
+
+            if not window:
+                continue
+            if self._shards:
+                close_waiter = asyncio.create_task(self._closing_event.wait())
+                shard_joiners = [
+                    asyncio.create_task(s.join(), name=f"wait for shard {s.id} to terminate")
+                    for s in self._shards.values()
+                ]
+
+                try:
+                    await aio.all_of(close_waiter, *shard_joiners, timeout=5)
+                    if close_waiter:
+                        _LOGGER.info("requested to shut down during startup of shards")
+                    else:
+                        _LOGGER.critical("one or more shards shut down unexpectedly during bot startup")
+                    return
+
+                except asyncio.TimeoutError:
+                    # new window starts.
+                    pass
+
+                except Exception as ex:
+                    _LOGGER.critical("an exception occurred in one of the started shards during bot startup: %r", ex)
+                    raise
+
+            started_shards = await aio.all_of(
+                *(
+                    self._start_one_shard(
+                        activity=activity,
+                        afk=afk,
+                        idle_since=idle_since,
+                        status=status,
+                        large_threshold=large_threshold,
+                        shard_id=candidate_shard_id,
+                        shard_count=shard_count,
+                        url=requirements.url,
+                    )
+                    for candidate_shard_id in window
+                    if candidate_shard_id in shard_ids
+                )
+            )
 
-        Returns
-        -------
-        builtins.float
-            The number of seconds the application has been running for.
-            If not started, then this will return `0.0`.
-        """
-        return date.monotonic() - self._started_at_monotonic if self._started_at_monotonic else 0.0
+            for started_shard in started_shards:
+                self._shards[started_shard.id] = started_shard
 
-    async def start(self) -> None:
-        """Determine sharding settings if needed, and start all shards.
+        await self.dispatch(lifetime_events.StartedEvent(app=self))
 
-        This method will first log if any library updates are available.
+    async def join(self, until_close: bool = True) -> None:
```

Comment:
What's the behaviour around this if join is called while it's in the middle of starting shards, will this leave some shards in an active state then?

Reply:
it will wait for any shards that are running.

We use it during windowing to detect if any shards have died while we back off for 5s.